src/server/search/global_hnsw_vector_index.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <absl/container/flat_hash_map.h>
+
+#include <memory>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include "core/search/base.h"
+#include "core/search/indices.h"
+#include "core/search/search.h"
+#include "server/tx_base.h"
+
+namespace dfly {
+
+struct KnnScoreSortOption;
+struct SearchResult;
+struct SearchParams;
+struct CommandContext;
+
+class GlobalHnswVectorIndex {
+ public:
+  GlobalHnswVectorIndex(const search::SchemaField::VectorParams& params,
+                        std::string_view index_name,
+                        PMR_NS::memory_resource* mr = PMR_NS::get_default_resource());
+
+  ~GlobalHnswVectorIndex();
+
+  bool Add(search::GlobalDocId id, const search::DocumentAccessor& doc, std::string_view field);
+  void Remove(search::GlobalDocId id, const search::DocumentAccessor& doc, std::string_view field);
+
+  std::vector<SearchResult> Search(
+      const search::AstKnnNode* knn,
+      const std::optional<search::KnnScoreSortOption>& knn_score_option,
+      const std::vector<SearchResult>& filter_docs, const SearchParams& params,
+      const CommandContext& cmd_cntx);
+
+ private:
+  std::vector<std::pair<float, search::GlobalDocId>> SearchKnnHnsw(
+      search::HnswVectorIndex* index, const search::AstKnnNode* knn,
+      const std::optional<std::vector<search::GlobalDocId>>& allowed_docs);
+
+  std::unique_ptr<search::BaseVectorIndex<search::GlobalDocId>> vector_index_;
```

Comment:
Probably it follows @dranikpg  question but what is the added value of declaring here BaseVectorIndex instead of specifying `search::HnswVectorIndex` explicitly. Moreover, what is the value of having  GlobalHnswVectorIndex wrappin `search::HnswVectorIndex` ? why not have only class?